cpu/stm32/periph/usbdev_fs.c

```diff
@@ -0,0 +1,671 @@
+
+/* Forward declaration for the usb device driver */
+const usbdev_driver_t driver;
+
+static USB_TypeDef *_global_regs(const stm32_usbdev_fs_config_t *conf)
+{
+    return (USB_TypeDef*)conf->base_addr;
+}
+
+/* Endpoint Buffer Descriptor */
+typedef struct {
+    uint16_t addr_tx;
+    uint16_t count_tx;
+    uint16_t addr_rx;
+    uint16_t count_rx;
+} ep_buf_desc_t;
+
+#define EP_DESC ((ep_buf_desc_t*)USB1_PMAADDR)
+#define EP_REG(x) (*((volatile uint32_t*) (USB1_BASE + (4*x))))
+
+usbdev_t *usbdev_get_ctx(unsigned num)
+{
+    assert(num < USBDEV_NUMOF);
+    return &_usbdevs[num].usbdev;
+}
+
+static inline void _enable_irq(void)
+{
+    stm32_usbdev_fs_t *usbdev = &_usbdevs[0];
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    _global_regs(conf)->CNTR = USB_CNTR_CTRM | USB_CNTR_RESETM  |
+                               USB_CNTR_ERRM | USB_CNTR_PMAOVRM;
+}
+
+static void _setup(stm32_usbdev_fs_t *usbdev,
+                      const stm32_usbdev_fs_config_t *config)
+{
+    usbdev->usbdev.driver = &driver;
+    usbdev->config = config;
+    usbdev->out = _ep_out;
+    usbdev->in = _ep_in;
+}
+
+static uint32_t _type_to_reg(usb_ep_type_t type)
+{
+    switch (type) {
+        case USB_EP_TYPE_CONTROL:
+            return USB_EP_CONTROL;
+        case USB_EP_TYPE_ISOCHRONOUS:
+            return USB_EP_ISOCHRONOUS;
+        case USB_EP_TYPE_BULK:
+            return USB_EP_BULK;
+        case USB_EP_TYPE_INTERRUPT:
+            return USB_EP_INTERRUPT;
+        default:
+            assert(false);
+            return 0;
+    }
+}
+
+void usbdev_init_lowlevel(void)
+{
+    for (size_t i = 0; i < USBDEV_NUMOF; i++) {
+        _setup(&_usbdevs[i], &stm32_usbdev_fs_config[i]);
+    }
+}
+
+static void _enable_usb_clk(void)
+{
+    /* Validate USB Supply */
+    PWR->CR2 |= PWR_CR2_USV;
+    RCC->APB1SMENR1 |= RCC_APB1SMENR1_USBSMEN;
+    /* Move it to specific init */
+    RCC->CSR |= RCC_CSR_RFRSTS;
+    RCC->CSR &= ~RCC_CSR_RFRSTS;
+    /* Enable CRS with auto trim enabled */
+    CRS->CR |=  CRS_CR_AUTOTRIMEN |  CRS_CR_CEN;
+}
+
+static void _enable_gpio(const stm32_usbdev_fs_config_t *conf)
+{
+    gpio_init(conf->dp, GPIO_IN);
+    gpio_init(conf->dm, GPIO_IN);
+    /* Configure AF for the pins */
+    gpio_init_af(conf->dp, conf->af);
+    gpio_init_af(conf->dm, conf->af);
+}
+
+static void _set_ep_in_status(uint16_t *val, uint16_t mask)
+{
+    /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPTX_DTOG1) {
+        *val ^= USB_EPTX_DTOG1;
+    } else {
+        if (*val & USB_EPTX_DTOG1) {
+            *val |= USB_EPTX_DTOG1;
+        }
+    }
+    if (mask & USB_EPTX_DTOG2) {
+        *val ^= USB_EPTX_DTOG2;
+    } else {
+        if (*val & USB_EPTX_DTOG2) {
+            *val |= USB_EPTX_DTOG2;
+        }
+    }
+}
+
+static void _set_ep_out_status(uint16_t *val, uint16_t mask)
```

Comment:
Is this function identical to the `_set_ep_in_status` function?

Reply:
No, USB_EPTX_DTOG1, USB_EPTX_DTOG2 are not the same as USB_EPRX_DTOG1, USB_EPRX_DTOG2

Reply:
I was looking for a tiny difference in the register names, but couldn't spot one. Thanks for clarifying.